From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paolo <paolo.sabatino@gmail.com>
Date: Thu, 22 Nov 2018 07:04:19 +0100
Subject: [ARCHEOLOGY] Add rk3288 xt-q8l-v10 CSC board (#1158)

> X-Git-Archeology: > recovered message: > This merge request contains various files which add support for xt-q8l-v10 boards (TVBox) equipped with Rockchip RK3288 SoC, AP6330 WiSoC (BCM4330 WiFi + Bluetooth), 2 GB DRAM (LPDDR2 or DDR3), 8 Gb eMMC, Gigabit Ethernet, 3 USB (1 OTG), 1 microSD slot, SPDIF optical output, 1 HDMI.
> X-Git-Archeology: > recovered message: > Kernel patches:
> X-Git-Archeology: > recovered message: > This thouches all three linux-rockchip-* kernelconfigs, just adds brcmfmac and brcmutil modules and remote controller support. default flavor activates rockchip own remote controller driver, next and dev use the mainline GPIO CIR driver (dev has lirc userland support activated too).
> X-Git-Archeology: > recovered message: > About the remote controller, an additional kernel module is added to the existing keymaps which is activated via device tree.
> X-Git-Archeology: > recovered message: > About possibly clashing patches assert-phy-reset-when-waking-up-in-rk3288-platform.patch should be checked against other rk3288 boards because it addresses an errata in rk3288 which causes the USB Host ports to stop responding when exiting from autosleep. On my device if I connect the first USB device when the system is already running, the USB Host gets stuck without this patch. Probably to work correctly on other platforms the device tree should include the proper reset lines of the USB PHYs (for reference, check patch/kernel/rockchip-dev/xt-q8l-v10-add-device-tree.patch starting from line 869).
> X-Git-Archeology: > recovered message: > Patch 1-2-regulator-act8865-add-restart-handler-for-act8846.patch adds a restart handler which allows reboot using SIPC bit on act8846 power regulator. Possibly MiQi board is affected (is reboot working there?), others (tinkerboard) should not care.
> X-Git-Archeology: > recovered message: > Patch brcmfmac-add-ap6330-firmware.patch adds firmware file names for ap6330 , should be harmless in other cases.
> X-Git-Archeology: > recovered message: > Patch 0010-GPU-Mali-Midgard-remove-rcu_read_lock-references.patch is from Miouyouyou. It should be harmless, it was suggested by him to do some tests with devfreq
> X-Git-Archeology: > recovered message: > Other patches just add the proper device trees, Kconfig and bits for supporting the board as a regular kernel supported board and should not interfere with anything else
> X-Git-Archeology: > recovered message: > U-Boot patches:
> X-Git-Archeology: > recovered message: > All the patches for u-boot are per-board, so nothing is added which may interfere with other existing boards here. They include the device tree and u-boot config and also a couple of patches to support the silergy power regulators driving current to CPU and GPU
> X-Git-Archeology: > recovered message: > * Initial commit to provide kernel and u-boot configuration and device trees for xt-q8-v10 as patches
> X-Git-Archeology: > recovered message: > Modification to rockchip config to add initialization bits for xt-q8-v10
> X-Git-Archeology: > recovered message: > * Committing correct path for rk3288_ddr_400Mhz... rockchip blob, moved assembling into another section to produce
> X-Git-Archeology: > recovered message: > immediately an u-boot working binary
> X-Git-Archeology: > recovered message: > * Enabled broadcom fmac driver in rockchip-next config
> X-Git-Archeology: > recovered message: > * Changed name definition of rk3288-xt-q8-v10 board to "TVBox"
> X-Git-Archeology: > recovered message: > Added bits to include support AP6330 and binary firmwares into the final image
> X-Git-Archeology: > recovered message: > * Fixed device tree file name in related patch, added patching of Makefile to produce the device tree binary accordingly
> X-Git-Archeology: > recovered message: > * Fixed xt-q8-v10 device tree patch
> X-Git-Archeology: > recovered message: > Added brcmfmac driver to rockchip dev and default kernel configs
> X-Git-Archeology: > recovered message: > * Syncing with upstream
> X-Git-Archeology: > recovered message: > * Splitted add-xt-q8... kernel patches into two separate patches
> X-Git-Archeology: > recovered message: > * Fixed bad extension while adding dtb in makefile for rockchip-default configuration
> X-Git-Archeology: > recovered message: > Updated device tree patches for all rockchip confs
> X-Git-Archeology: > recovered message: > * Enable mmc0 and usb in u-boot config
> X-Git-Archeology: > recovered message: > Fixed again makefile patch for kernel next
> X-Git-Archeology: > recovered message: > * Adding patches to reset the USB phy when kernel requires a reset, fixes autosuspend issue
> X-Git-Archeology: > recovered message: > * Changed xt-q8-v10 to proper xt-q8l-v10 in every string and every filename
> X-Git-Archeology: > recovered message: > Added power hold to u-boot, so now the device will boot and stay turned on without the need for the OTG cable anymore
> X-Git-Archeology: > recovered message: > * Changed names from 'Q8' to proper 'XT-Q8L-V10' in device tree patch files
> X-Git-Archeology: > recovered message: > * Legacy kernel device tree:
> X-Git-Archeology: > recovered message: > Fixed bluetooth gpio pin clashing
> X-Git-Archeology: > recovered message: > Fixed HDMI gpio pin clashing
> X-Git-Archeology: > recovered message: > Added support for PWM-based IR-Receiver, added driver in kernel default config too
> X-Git-Archeology: > recovered message: > Various other fixes to avoid some complaints from the kernel
> X-Git-Archeology: > recovered message: > * Added booting bluetooth systemd service for AP6330 (xt-q8l-v10) that loads patchram and invokes hciattach
> X-Git-Archeology: > recovered message: > Minor fixes to -next and -dev device trees for xt-q8l-v10
> X-Git-Archeology: > recovered message: > * Disabled OTG USB port in u-boot due to long timeout during initialization
> X-Git-Archeology: > recovered message: > Fixed warning during u-boot dts compilation
> X-Git-Archeology: > recovered message: > Added emmc as second boot device in dts
> X-Git-Archeology: > recovered message: > * Adding myself to licensing
> X-Git-Archeology: > recovered message: > * Committing modifications to device trees
> X-Git-Archeology: > recovered message: > * Fixed dmac_bus_s explicitly set to unused dmac, restored right dmac in xt-q8l-v10 dts only
> X-Git-Archeology: > recovered message: > Change PLL_CPLL frequency in device tree to 408 Mhz to avoid fractional divisor warnings
> X-Git-Archeology: > recovered message: > * Added proper xt-q8l-v10_rk3288 configuration to u-boot, now appearing in config menu and
> X-Git-Archeology: > recovered message: > correctly selectable as a real target
> X-Git-Archeology: > recovered message: > Fixed typo in device tree from rockchip
> X-Git-Archeology: > recovered message: > * Fixed missing semicolon in device tree for default configuration
> X-Git-Archeology: > recovered message: > Fixed patch files for u-boot appending themselves to files on each compilation
> X-Git-Archeology: > recovered message: > * Added bits to enable power to USB ports in u-boot, thus enabling booting from USB devices (only USB host port for now)
> X-Git-Archeology: > recovered message: > * Changed u-boot binary creation using the rockchip SPL properly
> X-Git-Archeology: > recovered message: > * Added boot order for xt-q8l-v10: sdcard, usb0, eMMC, network
> X-Git-Archeology: > recovered message: > * Added bionic:next in beta config for xt-q8l-v10 board
> X-Git-Archeology: > recovered message: > * Changed some minor bits in xt-q8l-v10 device tree files, added missing bits to dev flavour
> X-Git-Archeology: > recovered message: > Added patches to introduce fairchild fan53555/silergy82x regulators to u-boot and enabled in xt-q8l-v10 device tree
> X-Git-Archeology: > recovered message: > * Updated u-boot to version v2018.03 for xt-q8l-v10. Other rk3288 boards will gain v2018.05 from main armbian fork
> X-Git-Archeology: > recovered message: > Removed pre-reloc labels in u-boot device tree because they are not necessary since we don't use u-boot SPL for xt-q8l-v10
> X-Git-Archeology: > recovered message: > Removed vmmc-supply and vqmmc-supply in u-boot device tree to avoid hang on boot
> X-Git-Archeology: > recovered message: > * Tidied up a bit device trees, in particular some modifications are made to power regulator properties comparing them against the original q8l device tree
> X-Git-Archeology: > recovered message: > Removed unnecessary dummy regulator, removed unnecessary capacities to embedded eMMC
> X-Git-Archeology: > recovered message: > Disabled unused USB host
> X-Git-Archeology: > recovered message: > Removed vmmc-supply and vqmmc-supply from emmc section because it causes hang in u-boot v2018.03 and newer
> X-Git-Archeology: > recovered message: > * Restored previous regulator in u-boot dts
> X-Git-Archeology: > recovered message: > removed assert phy reset USB patch from rockchip-dev because of some upstream incompatible changes
> X-Git-Archeology: > recovered message: > * Added patch to enable IRQ for Midgard drivers which caused massive slowdown on dev kernel
> X-Git-Archeology: > recovered message: > Changed u-boot if-code for xt-q8l-v10 in rockchip.conf
> X-Git-Archeology: > recovered message: > Removed references to rk3288-linux.dtsi in xt-q8l-v10 device tree for default kernel
> X-Git-Archeology: > recovered message: > * Committing effective removal of USB reset assert for dev kernel
> X-Git-Archeology: > recovered message: > Committing changes to u-boot device tree
> X-Git-Archeology: > recovered message: > * Added patch to disable USB power down for rockchip devices broken on latest kernel
> X-Git-Archeology: > recovered message: > * Removed usb dwc2 patch to reinject it from specific branch
> X-Git-Archeology: > recovered message: > * Reverting some voltage changes for xt-q8l-v10 device in rockchip-dev
> X-Git-Archeology: > recovered message: > * Reverting some voltage changes for xt-q8l-v10 in u-boot section
> X-Git-Archeology: > recovered message: > * Added patch to make USB ports working again on rockchip devices with mainline
> X-Git-Archeology: > recovered message: > kernel >= 4.18
> X-Git-Archeology: > recovered message: > * Changed the 0 into false
> X-Git-Archeology: > recovered message: > * Moved xt-q8l-v10 u-boot patches into board_xt-q8l-v10 directory
> X-Git-Archeology: > recovered message: > * Changed some minor things in rockchip-dev dts for xt-q8l-v10, added mali midgard driver to dev kernel config
> X-Git-Archeology: > recovered message: > * Added devfreq support for Mali in rockchip-next flavour
> X-Git-Archeology: > recovered message: > * Remove manually applied patch (0007-drivers-drm...) because it has been
> X-Git-Archeology: > recovered message: > added to armbian main repo
> X-Git-Archeology: > recovered message: > * Removed duplicate patch which has added to main armbian repository
> X-Git-Archeology: > recovered message: > * Tidied up regulators for default/next/dev rockchip flavours for xt-q8l-v10, disabling those regulators which are not tied to anything
> X-Git-Archeology: > recovered message: > Enabled voltage regulator to make SPDIF connector work (thus not tested because I have no DAC)
> X-Git-Archeology: > recovered message: > Changed rockchip-dev and rockchip-next config files to enable gpio-ir-receiver module to enable bundled remote IR controller, including kernel patch for keymap
> X-Git-Archeology: > recovered message: > * Enabled back regulator REG7 to allow propert bluetooth functionaly
> X-Git-Archeology: > recovered message: > * Minor changes to u-boot device tree for xt-q8l-v10
> X-Git-Archeology: > recovered message: > Added patch to set act8846 SIPC to correctly reboot the device (thus require some power-hold at reboot to make reboot fully working)
> X-Git-Archeology: > recovered message: > * Fixed u-boot device tree
> X-Git-Archeology: > recovered message: > * Added configuration bits to support TPL in u-boot for xt-q8l-v10 (TPL is thrown away though) to allow faster reboot times and achieve a working reset feature activating power hold gpio pin as soon as possible. gpio pin is hardwired into spl_board_init() u-boot code because it is not possible to let it work via device tree
> X-Git-Archeology: > recovered message: > Fixed OTG USB port in u-boot, allowing devices detection and booting
> X-Git-Archeology: > recovered message: > Added proper vbus-supply properties for USB controllers in u-boot dts, so u-boot activates USB vbus itself
> X-Git-Archeology: > recovered message: > * Fixed dts makefile patching for next and dev rockchip kernel
> X-Git-Archeology: > recovered message: > * Fixed fdt_file renamed to fdtfile in armbianEnv.txt
> X-Git-Archeology: > recovered message: > * Changed xt-q8l-v10 board config as per recomendations
> X-Git-Archeology: > recovered message: > * Moved xt-q8l-v10 configuration to CSC
> X-Git-Archeology: > recovered message: > Restored linux-rockchip-* configurations, enabled brcmfmac driver, GPIO remote controller driver and lirc kernel compatibility interface
> X-Git-Archeology: > recovered message: > Polished a bit rockchip.conf
> X-Git-Archeology: > recovered message: > * Add patch to brcmfmac driver to search for ap6330 firmware
> X-Git-Archeology: > recovered message: > Removed copy-work from rockchip.conf about ap6330 firmware for xt-q8l-v10 and tidied up
> X-Git-Archeology: > recovered message: > Avoid using brcm_patchram_plus in ap6330-bluetooth-service putting proper firmware file in /etc/firmware for hciattach do firmware uploading itself
> X-Git-Archeology: > recovered message: > * Fixed bcm4330 bluetooth firmware linking for hciattach used by ap6330-bluetooth.service
> X-Git-Archeology: > recovered message: > * Removed foreign test patches from xt-q8l-v10 u-boot directory
> X-Git-Archeology: - Revision 60b4166a8a9efe74c76bf75246cd297ccf4cf7ca: https://github.com/armbian/build/commit/60b4166a8a9efe74c76bf75246cd297ccf4cf7ca
> X-Git-Archeology:   Date: Thu, 22 Nov 2018 07:04:19 +0100
> X-Git-Archeology:   From: Paolo <paolo.sabatino@gmail.com>
> X-Git-Archeology:   Subject: Add rk3288 xt-q8l-v10 CSC board (#1158)
> X-Git-Archeology:
> X-Git-Archeology: - Revision f4cce9754879f1d8e956b5ee7dc05b6d049f0e94: https://github.com/armbian/build/commit/f4cce9754879f1d8e956b5ee7dc05b6d049f0e94
> X-Git-Archeology:   Date: Wed, 10 Jun 2020 20:35:52 +0200
> X-Git-Archeology:   From: Paolo <paolo.sabatino@gmail.com>
> X-Git-Archeology:   Subject: [rk3288] Various fixes and enhancements for xt-q8l-v10 CSC board (#2013)
> X-Git-Archeology:
> X-Git-Archeology: - Revision de26797423e22d58ec2882d7032c67f77196ecc5: https://github.com/armbian/build/commit/de26797423e22d58ec2882d7032c67f77196ecc5
> X-Git-Archeology:   Date: Sun, 06 Nov 2022 20:32:46 +0100
> X-Git-Archeology:   From: Igor Pecovnik <igorpecovnik@users.noreply.github.com>
> X-Git-Archeology:   Subject: Move all legacy u-boot patches under one general legacy folder (#4386)
> X-Git-Archeology:
> X-Git-Archeology: - Revision 97c34489831f2146940f52915428263b7edfcbe1: https://github.com/armbian/build/commit/97c34489831f2146940f52915428263b7edfcbe1
> X-Git-Archeology:   Date: Fri, 24 Mar 2023 23:13:42 +0100
> X-Git-Archeology:   From: Paolo Sabatino <paolo.sabatino@gmail.com>
> X-Git-Archeology:   Subject: rockchip: put all rockchip 32 bit into uboot/v2022.04 directory
> X-Git-Archeology:
---
 arch/arm/dts/Makefile                                |  1 +
 arch/arm/mach-rockchip/rk3288/Kconfig                | 12 ++++
 board/rockchip/xt-q8l-v10_rk3288/Kconfig             | 15 +++++
 board/rockchip/xt-q8l-v10_rk3288/MAINTAINERS         |  6 ++
 board/rockchip/xt-q8l-v10_rk3288/Makefile            |  7 +++
 board/rockchip/xt-q8l-v10_rk3288/xt-q8l-v10-rk3288.c | 21 +++++++
 include/configs/xt-q8l-v10_rk3288.h                  | 28 ++++++++++
 7 files changed, 90 insertions(+)

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 111111111111..222222222222 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -109,6 +109,7 @@ dtb-$(CONFIG_ROCKCHIP_RK3288) += \
 	rk3288-rock-pi-n8.dtb \
 	rk3288-tinker.dtb \
 	rk3288-tinker-s.dtb \
+	rk3288-xt-q8l-v10.dtb \
 	rk3288-veyron-jerry.dtb \
 	rk3288-veyron-mickey.dtb \
 	rk3288-veyron-minnie.dtb \
diff --git a/arch/arm/mach-rockchip/rk3288/Kconfig b/arch/arm/mach-rockchip/rk3288/Kconfig
index 111111111111..222222222222 100644
--- a/arch/arm/mach-rockchip/rk3288/Kconfig
+++ b/arch/arm/mach-rockchip/rk3288/Kconfig
@@ -134,6 +134,16 @@ config TARGET_TINKER_RK3288
 	  8GB eMMC and 2GB of SDRAM. Expansion connectors provide access to
 	  I2C, SPI, UART, GPIOs.
 
+config TARGET_XT_Q8L_V10_RK3288
+       bool "xt-q8l-v10 tvbox board"
+       select BOARD_LATE_INIT
+       select TPL
+       help
+         xt-q8l-v10 is a RK3288 common tv box with 3 USB ports (1 OTG), HDMI,
+         micro-SD card, Gigabit Ethernet and SPDIF output. It also includes on-board
+         8GB eMMC and 2GB of SDRAM, 802.11n Wifi and Bluetooth based upon AP6330
+         WiSoC.
+
 endchoice
 
 config ROCKCHIP_FAST_SPL
@@ -196,4 +206,6 @@ source "board/rockchip/evb_rk3288/Kconfig"
 
 source "board/rockchip/tinker_rk3288/Kconfig"
 
+source "board/rockchip/xt-q8l-v10_rk3288/Kconfig"
+
 endif
diff --git a/board/rockchip/xt-q8l-v10_rk3288/Kconfig b/board/rockchip/xt-q8l-v10_rk3288/Kconfig
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/board/rockchip/xt-q8l-v10_rk3288/Kconfig
@@ -0,0 +1,15 @@
+if TARGET_XT_Q8L_V10_RK3288
+
+config SYS_BOARD
+	default "xt-q8l-v10_rk3288"
+
+config SYS_VENDOR
+	default "rockchip"
+
+config SYS_CONFIG_NAME
+	default "xt-q8l-v10_rk3288"
+
+config BOARD_SPECIFIC_OPTIONS # dummy
+	def_bool y
+
+endif
diff --git a/board/rockchip/xt-q8l-v10_rk3288/MAINTAINERS b/board/rockchip/xt-q8l-v10_rk3288/MAINTAINERS
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/board/rockchip/xt-q8l-v10_rk3288/MAINTAINERS
@@ -0,0 +1,6 @@
+XT-Q8L-V10-RK3288
+M:	Paolo Sabatino <a@b.c>
+S:	Not mainlined
+F:	board/rockchip/xt-q8l-v10_rk3288
+F:	include/configs/xt-q8l-v10_rk3288.h
+F:	configs/xt-q8l-v10-rk3288_defconfig
diff --git a/board/rockchip/xt-q8l-v10_rk3288/Makefile b/board/rockchip/xt-q8l-v10_rk3288/Makefile
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/board/rockchip/xt-q8l-v10_rk3288/Makefile
@@ -0,0 +1,7 @@
+#
+# (C) Copyright 2016 Rockchip Electronics Co., Ltd
+#
+# SPDX-License-Identifier:     GPL-2.0+
+#
+
+obj-y	+= xt-q8l-v10-rk3288.o
diff --git a/board/rockchip/xt-q8l-v10_rk3288/xt-q8l-v10-rk3288.c b/board/rockchip/xt-q8l-v10_rk3288/xt-q8l-v10-rk3288.c
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/board/rockchip/xt-q8l-v10_rk3288/xt-q8l-v10-rk3288.c
@@ -0,0 +1,21 @@
+/*
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <common.h>
+#include <spl.h>
+#include <asm/gpio.h>
+
+void spl_board_init(void)
+{
+
+	/*
+        * xt-q8l-v10 requires the power-hold gpio set (A11, or bank 0 pin 11
+        * if you prefer) as soon as possible.
+        * We hack it here :P
+        */
+       gpio_request(11, "power-hold");
+       gpio_direction_output(11, 1);
+
+}
+
diff --git a/include/configs/xt-q8l-v10_rk3288.h b/include/configs/xt-q8l-v10_rk3288.h
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/include/configs/xt-q8l-v10_rk3288.h
@@ -0,0 +1,28 @@
+/*
+ * (C) Copyright 2015 Google, Inc
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#ifndef __CONFIG_H
+#define __CONFIG_H
+
+#define ROCKCHIP_DEVICE_SETTINGS \
+		"stdin=serial,usbkbd\0" \
+		"stdout=serial,vidconsole\0" \
+		"stderr=serial,vidconsole\0"
+
+#include <configs/rk3288_common.h>
+
+#undef BOOT_TARGET_DEVICES
+
+#define BOOT_TARGET_DEVICES(func) \
+        func(MMC, mmc, 1) \
+        func(USB, usb, 0) \
+	func(MMC, mmc, 0) \
+        func(PXE, pxe, na) \
+        func(DHCP, dchp, na)
+
+#define CONFIG_SYS_MMC_ENV_DEV 0
+
+#endif
-- 
Armbian

